connexion/apps/async_app.py

```diff
@@ -0,0 +1,268 @@
+"""
+This module defines a native connexion asynchronous application.
+"""
+
+import asyncio
+import logging
+import pathlib
+import pkgutil
+import sys
+import typing as t
+
+from starlette.responses import JSONResponse as StarletteJSONResponse
+from starlette.responses import Response as StarletteResponse
+from starlette.routing import Router
+from starlette.types import Receive, Scope, Send
+
+from connexion.apis.abstract import AbstractAPI
+from connexion.apps.abstract import AbstractApp
+from connexion.decorators.lifecycle import RequestResponseDecorator
+from connexion.decorators.parameter import parameter_to_arg
+from connexion.exceptions import MissingMiddleware, ProblemException
+from connexion.lifecycle import ConnexionResponse, MiddlewareRequest, MiddlewareResponse
+from connexion.middleware.main import ConnexionMiddleware
+from connexion.middleware.routing import ROUTING_CONTEXT
+from connexion.operations import AbstractOperation
+from connexion.uri_parsing import AbstractURIParser
+
+logger = logging.getLogger("Connexion.app")
+
+
+class AsyncAsgiApp:
+    """Mixin for usage with Abstract Apps & Apis"""
+
+    def __init__(self, *args, base_path="", **kwargs):
+        self.apis: t.Dict[str, AsyncApi] = {}
+        self.operations: t.Dict[str, AsyncOperation] = {}
+        self.router = Router()
+        self.base_path = base_path
+        super(AsyncAsgiApp, self).__init__(*args, **kwargs)
+
+    def add_api(self, specification: t.Union[pathlib.Path, str, dict], **kwargs):
+        api = super(AsyncAsgiApp, self).add_api(specification, **kwargs)  # type: ignore
+        self.apis[api.base_path] = api
+        return api
+
+    def add_url_rule(self, rule, endpoint=None, view_func=None, **options):
+        self.router.add_route(path=rule, endpoint=view_func, name=endpoint, **options)
+
+    def route(self, rule, **options):
+        self.router.route(rule, **options)
+
+    async def asgi_app(self, scope: Scope, receive: Receive, send: Send) -> None:
+        """The actual ASGI application. This is not implemented in
+        `__call__` so that middlewares can be applied without
+        losing a reference to the app object.
+        """
+        try:
+            connexion_context = scope["extensions"][ROUTING_CONTEXT]
+        except KeyError:
+            raise MissingMiddleware(
+                "Could not find routing information in scope. Please make sure "
+                "you have a routing middleware registered upstream. "
+            )
+
+        api_base_path = connexion_context.get("api_base_path")
+        if api_base_path and not api_base_path == self.base_path:
```

Comment:
We might be able to leverage the RoutedMiddleware base classes here, will try to rework.